src/upgrader.js

```diff
@@ -0,0 +1,359 @@
+        const { stream, protocol } = await mss.handle(Array.from(this.protocols.keys()))
+        log('outbound: incoming stream opened on %s', protocol)
+        connection.addStream(stream, protocol)
+        this._onStream({ stream, protocol })
+      },
+      // Run anytime a stream closes
+      onStreamEnd: muxedStream => {
+        connection.removeStream(muxedStream.id)
+      }
+    })
+
+    const newStream = async protocols => {
+      log('outbound: starting new stream on %s', protocols)
+      const muxedStream = muxer.newStream()
+      const mss = new Multistream.Dialer(muxedStream)
+      try {
+        const { stream, protocol } = await mss.select(protocols)
+        return { stream: { ...muxedStream, ...stream }, protocol }
+      } catch (err) {
+        log.error('could not create new stream', err)
+        throw errCode(err, codes.ERR_UNSUPPORTED_PROTOCOL)
+      }
+    }
+
+    // Pipe all data through the muxer
+    pipe(muxedConnection, muxer, muxedConnection)
+
+    maConn.timeline.upgraded = Date.now()
+    const timelineProxy = new Proxy(maConn.timeline, {
+      set: (...args) => {
+        if (args[1] === 'close' && args[2]) {
+          this.onConnectionEnd(connection)
+        }
+
+        return Reflect.set(...args)
+      }
+    })
+
+    // Create the connection
+    const connection = new Connection({
+      localAddr: maConn.localAddr,
+      remoteAddr: maConn.remoteAddr,
+      localPeer: this.localPeer,
+      remotePeer: remotePeer,
+      stat: {
+        direction: 'outbound',
+        timeline: timelineProxy,
+        multiplexer: Muxer.multicodec,
+        encryption: cryptoProtocol
+      },
+      newStream,
+      getStreams: () => muxer.streams,
+      close: err => maConn.close(err)
+    })
+
+    this.onConnection(connection)
+
+    return connection
+  }
+
+  /**
+   * Routes incoming streams to the correct handler
+   * @private
+   * @param {object} options
+   * @param {Stream} options.stream
+   * @param {string} protocol
+   */
+  _onStream ({ stream, protocol }) {
+    const handler = this.protocols.get(protocol)
+    handler({ stream, protocol })
+  }
+
+  /**
+   * Attempts to encrypt the incoming `connection` with the provided `cryptos`.
+   * @private
+   * @async
+   * @param {PeerId} localPeer The initiators PeerInfo
+   * @param {*} connection
+   * @param {Map<string, Crypto>} cryptos
+   * @returns {[connection, string]} An encrypted connection and the tag of the `Crypto` used
+   */
+  async _encryptInbound (localPeer, connection, cryptos) {
+    const mss = new Multistream.Listener(connection)
+    const protocols = Array.from(cryptos.keys())
+    log('selecting inbound crypto protocol', protocols)
+
+    try {
+      const { stream, protocol } = await mss.handle(protocols)
+      const crypto = cryptos.get(protocol)
+      log('encrypting inbound connection...')
+      const cryptoResponse = await crypto.secureInbound(localPeer, stream)
+
+      if (cryptoResponse) {
```

Comment:
What happens if we do not return?

Reply:
This should just throw, there's no need to check the result, fixing this.